File: python/107_BinaryTreeLevelOrderTraversalII.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        levels = {}
        self.deepest_level = 0
        def traverseTree(node, lvl):
            if not node:
                return None
            self.deepest_level = max(self.deepest_level, lvl)
            levels[lvl] = levels[lvl] + [node.val] if lvl in levels else [node.val]
            
            return traverseTree(node.left, lvl+1) or traverseTree(node.right, lvl+1)
        
        traverseTree(root, 0)
        return [levels[lvl] for lvl in range(self.deepest_level,-1,-1)] 

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        self.levels = 0
        def countLevels(node, level):
            if not node:
                return self.levels
            
            if level > self.levels:
                self.levels = level
            
            if node.left:
                countLevels(node.left, level+1)
            if node.right:
                countLevels(node.right, level+1)
            return self.levels
        
        deepest_level = countLevels(root, 1)
        res = [[] for _ in range(deepest_level)]
        
        def helper(node, level):
            if not node:
                return None
            
            res[len(res)-level] += [node.val]
            if node.left:
                helper(node.left, level+1)
            if node.right:
                helper(node.right, level+1)
                
            return res
            
        return helper(root, 1)

File: java/107_BinaryTreeLevelOrderTraversalII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Integer, List<Integer>> levels;
    int deepestLevel;
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        if (root == null) return new ArrayList<List<Integer>>();
        
        levels = new HashMap<Integer, List<Integer>>();
        traverseTree(root, 0);
        
        List<List<Integer>> levelList = new ArrayList<List<Integer>>();
        for (int i = deepestLevel; i >= 0; --i) {
            levelList.add(levels.get(i));
        }
        return levelList;
        
    }
    
    public boolean traverseTree(TreeNode node, int lvl) {
        if (node == null) return false;
        
        deepestLevel = lvl > deepestLevel ? lvl : deepestLevel;
        
        // HashMap.containsKey returns null if doesn't contain key and you can only compare null != true
        if (levels.containsKey(lvl) != true) {
            levels.put(lvl, new ArrayList<Integer>(Arrays.asList(node.val)));
        } else {
            levels.get(lvl).add(node.val);
        }
        return traverseTree(node.left, lvl+1) || traverseTree(node.right, lvl+1);
    }
}
